src/runtime_src/core/edge/drm/zocl/sched_exec.c

```diff
@@ -911,6 +915,65 @@ configure(struct sched_cmd *cmd)
 	return 0;
 }
 
+/**
+ * Gather ERT stats in ctrl command packet
+ * [1  ]      : header
+ * [1  ]      : custat version
+ * [1  ]      : ert version
+ * [1  ]      : number of cq slots
+ * [1  ]      : number of cus
+ * [#numcus]  : cu execution stats (number of executions)
+ * [#numcus]  : cu status (1: running, 0: idle)
+ * [#slots]   : command queue slot status
+ */
+static void
+cu_stat(struct sched_cmd *cmd)
+{
+	struct drm_zocl_dev *zdev = cmd->ddev->dev_private;
+	struct zocl_ert_dev *ert = zdev->ert;
+	struct sched_exec_core *exec = zdev->exec;
+	struct ert_packet *pkg;
+	int slot_idx;
+	int pkt_idx = 0;
+	int max_idx = (slot_size(cmd->ddev) >> 2) - 1;
+	int i;
+
+	SCHED_DEBUG("-> %s cq_slot_idx %d\n", __func__, cmd->cq_slot_idx);
+	slot_idx = cmd->cq_slot_idx;
+	pkg = cmd->packet;
+
+	/* custat version, update when changing layout of packet */
+	pkg->data[pkt_idx++] = 0x51a10000;
+
+	/* should be git version */
+	pkg->data[pkt_idx++] = ERT_VERSION;
+
+	/* number of CQ slots */
+	pkg->data[pkt_idx++] = exec->num_slots;
+
+	/* number of CUs */
+	pkg->data[pkt_idx++] = exec->num_cus;
+
+	/* individual CU execution stat */
+	for (i = 0; i < exec->num_cus && pkt_idx < max_idx; ++i)
+		pkg->data[pkt_idx++] = exec->zcu[i].usage;
+
+	/* individual CU status */
+	for (i = 0; i < exec->num_cus && pkt_idx < max_idx; ++i)
+		pkg->data[pkt_idx++] = exec->cu_status[i];
+
+	/* Command slot status
+	 * Hard code QUEUED state. When a NEW command is found,
+	 * ERT/PS will set it to QUEUED. Then no CQ write.
+	 */
+	for (i = 0; i < exec->num_slots && pkt_idx < max_idx; ++i)
+		pkg->data[pkt_idx++] = ERT_CMD_STATE_QUEUED;
```

Comment:
why can't we return the real slot status with idle or queued?

Reply:
ERT/PS will copy command from CQ to PS memory to avoid AXI Lite access to the PL. If the real slot is CQ, the status is queued. If the "real slot" is the copy in PS memory, then we need to iterate  queued command list(maybe pending list also) and update corresponding CQ slot. I just too lazy to implement it for now. Let's refine this in KDS/ERT 3.0.

Reply:
I was thinking to just check the cmd state in each CQ slot just like what we are doing in interate_packets. But I am ok to refine it later.